mesa/agent.py

```diff
@@ -161,23 +160,35 @@ def select(
             AgentSet: A new AgentSet containing the selected agents, unless inplace is True, in which case the current AgentSet is updated.
         """
 
-        def agent_generator():
+        if filter_func is None and agent_type is None and n == 0:
+            return self if inplace else copy.copy(self)
+
+        def agent_generator(filter_func=None, agent_type=None, n=0):
             count = 0
             for agent in self:
-                if filter_func and not filter_func(agent):
-                    continue
-                if agent_type and not isinstance(agent, agent_type):
-                    continue
-                yield agent
-                count += 1
-                # default of n is zero, zo evaluates to False
-                if n and count >= n:
-                    break
+                if (not filter_func or filter_func(agent)) and (
+                    not agent_type or isinstance(agent, agent_type)
+                ):
+                    yield agent
+                    count += 1
+                    if 0 < n <= count:
+                        break
 
-        agents = agent_generator()
+        agents = agent_generator(filter_func, agent_type, n)
 
         return AgentSet(agents, self.model) if not inplace else self._update(agents)
 
+    def _agent_generator(self, filter_func=None, agent_type=None, n=0):
```

Comment:
This seems to be identical to `agent_generator`. And if there is a way to deduplicate the definition.

Reply:
whoopsie, you are correct, this shouldn't be here anymore. Removed it now